File: hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/functional/TestClusteringWithCustomMerger.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.hudi.client.functional;

import org.apache.hudi.DefaultSparkRecordMerger;
import org.apache.hudi.client.SparkRDDWriteClient;
import org.apache.hudi.client.WriteClientTestUtils;
import org.apache.hudi.common.config.HoodieStorageConfig;
import org.apache.hudi.common.config.RecordMergeMode;
import org.apache.hudi.common.model.HoodieBaseFile;
import org.apache.hudi.common.model.HoodieRecord;
import org.apache.hudi.common.model.HoodieRecordMerger;
import org.apache.hudi.common.model.HoodieTableType;
import org.apache.hudi.common.table.HoodieTableMetaClient;
import org.apache.hudi.common.table.timeline.HoodieInstant;
import org.apache.hudi.common.table.timeline.HoodieTimeline;
import org.apache.hudi.common.table.view.FileSystemViewStorageConfig;
import org.apache.hudi.common.testutils.HoodieTestDataGenerator;
import org.apache.hudi.common.util.Option;
import org.apache.hudi.config.HoodieClusteringConfig;
import org.apache.hudi.config.HoodieCompactionConfig;
import org.apache.hudi.config.HoodieIndexConfig;
import org.apache.hudi.config.HoodieWriteConfig;
import org.apache.hudi.index.HoodieIndex;
import org.apache.hudi.testutils.HoodieClientTestUtils;
import org.apache.hudi.testutils.SparkClientFunctionalTestHarness;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Properties;
import java.util.stream.Stream;

import static org.apache.hudi.common.testutils.HoodieTestDataGenerator.TRIP_EXAMPLE_SCHEMA;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Reproduction for HUDI issue #18980: standalone/inline clustering on a MOR table configured with
 * {@code hoodie.write.record.merge.mode=CUSTOM} fails with
 * "No valid spark merger implementation set for `hoodie.write.record.merge.custom.implementation.classes`".
 *
 * <p>The failing path is the RDD-based clustering execution
 * ({@code MultipleSparkJobExecutionStrategy} -> {@code ClusteringExecutionStrategy.getFileGroupReader}),
 * so row writer is disabled to exercise it.
 */
@Tag("functional")
class TestClusteringWithCustomMerger extends SparkClientFunctionalTestHarness {

  public static class CustomMerger extends DefaultSparkRecordMerger {
    @Override
    public String getMergingStrategy() {
      return HoodieRecordMerger.CUSTOM_MERGE_STRATEGY_UUID;
    }
  }

  @Test
  void clusteringWithCustomMergerShouldSucceed() throws Exception {
    HoodieWriteConfig cfg = getHoodieWriteConfig(false);
    HoodieTableMetaClient metaClient = getHoodieMetaClient(HoodieTableType.MERGE_ON_READ, cfg.getProps());
    HoodieTestDataGenerator dataGen = new HoodieTestDataGenerator();

    try (SparkRDDWriteClient client = getHoodieWriteClient(cfg)) {
      String commit1 = "001";
      WriteClientTestUtils.startCommitWithTime(client, commit1);
      List<HoodieRecord> records = dataGen.generateInserts(commit1, 400);
      Stream<HoodieBaseFile> dataFiles =
          insertRecordsToMORTable(metaClient, records.subList(0, 200), client, cfg, commit1);
      assertTrue(dataFiles.findAny().isPresent(), "should list the base files we wrote in the delta commit");

      String commit2 = "002";
      WriteClientTestUtils.startCommitWithTime(client, commit2);
      dataFiles = insertRecordsToMORTable(metaClient, records.subList(200, 400), client, cfg, commit2);
      assertTrue(dataFiles.findAny().isPresent(), "should list the base files we wrote in the delta commit");

      String commit3 = "003";
      WriteClientTestUtils.startCommitWithTime(client, commit3);
      updateRecordsInMORTable(metaClient, dataGen.generateUpdates(commit3, 100), client, cfg, commit3, false);
    }

    HoodieWriteConfig cfgClustering = getHoodieWriteConfig(true);
    String clusteringCommitTime;
    try (SparkRDDWriteClient client = getHoodieWriteClient(cfgClustering)) {
      clusteringCommitTime = client.scheduleClustering(Option.empty()).get().toString();
      // Issue #18980 surfaces here: the clustering execution strategy reads the source file groups
      // through HoodieFileGroupReader but does not propagate the custom merger impl classes, so
      // initRecordMerger throws "No valid spark merger implementation set".
      client.cluster(clusteringCommitTime, true);
    }

    // Clustering must have produced a completed replace commit at the scheduled instant.
    metaClient = HoodieTableMetaClient.reload(metaClient);
    HoodieTimeline timeline = metaClient.getCommitTimeline().filterCompletedInstants();
    HoodieInstant lastInstant = timeline.lastInstant().get();
    assertEquals(HoodieTimeline.REPLACE_COMMIT_ACTION, lastInstant.getAction(),
        "Clustering should complete as a replace commit");
    assertEquals(clusteringCommitTime, lastInstant.requestedTime(),
        "The last commit should be the clustering instant");
    // And all 400 distinct records must still be readable after clustering.
    assertEquals(400, HoodieClientTestUtils.countRecordsOptionallySince(
        jsc(), basePath(), sqlContext(), timeline, Option.empty()),
        "All records should be readable after clustering");
  }

  private HoodieWriteConfig getHoodieWriteConfig(boolean enableInlineClustering) {
    Properties props = new Properties();
    // Force the RDD-based clustering execution path (the one that reproduces the bug).
    props.setProperty("hoodie.datasource.write.row.writer.enable", "false");
    return HoodieWriteConfig.newBuilder()
        .forTable("test-custom-merger-clustering")
        .withPath(basePath())
        .withSchema(TRIP_EXAMPLE_SCHEMA)
        .withParallelism(2, 2)
        .withDeleteParallelism(2)
        .withProps(props)
        .withPreCombineField("timestamp")
        .withRecordMergeMode(RecordMergeMode.CUSTOM)
        .withRecordMergeImplClasses(CustomMerger.class.getName())
        .withRecordMergeStrategyId(HoodieRecordMerger.CUSTOM_MERGE_STRATEGY_UUID)
        .withCompactionConfig(HoodieCompactionConfig.newBuilder()
            .compactionSmallFileSize(10L)
            .withInlineCompaction(false).withMaxNumDeltaCommitsBeforeCompaction(1).build())
        .withStorageConfig(HoodieStorageConfig.newBuilder()
            .hfileMaxFileSize(1024 * 1024 * 1024)
            .parquetMaxFileSize(1024 * 1024 * 1024).build())
        .withEmbeddedTimelineServerEnabled(true)
        .withFileSystemViewConfig(new FileSystemViewStorageConfig.Builder()
            .withEnableBackupForRemoteFileSystemView(false).build())
        .withIndexConfig(HoodieIndexConfig.newBuilder().withIndexType(HoodieIndex.IndexType.BLOOM).build())
        .withClusteringConfig(HoodieClusteringConfig.newBuilder()
            .withClusteringMaxNumGroups(10)
            .withClusteringTargetPartitions(0)
            .withInlineClustering(enableInlineClustering)
            .withInlineClusteringNumCommits(1)
            .build())
        .withRollbackUsingMarkers(false)
        .build();
  }
}

File: hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/procedure/TestClusteringWithCustomMerger.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.spark.sql.hudi.procedure

import org.apache.hudi.{DefaultSparkRecordMerger, HoodieDataSourceHelpers}
import org.apache.hudi.common.config.HoodieReaderConfig
import org.apache.hudi.common.model.HoodieRecordMerger
import org.apache.hudi.common.table.timeline.HoodieTimeline

import org.apache.hadoop.fs.Path

import scala.collection.JavaConverters._

/**
 * Regression test for <a href="https://github.com/apache/hudi/issues/18980">HUDI issue #18980</a>:
 * clustering on a table configured with {@code hoodie.write.record.merge.mode=CUSTOM} and a custom
 * Spark merger failed with
 * "No valid spark merger implementation set for `hoodie.write.record.merge.custom.implementation.classes`".
 *
 * <p>The merge mode and strategy id are persisted as table config, but the custom merger impl classes
 * are a write-side config that is not persisted. Before the fix,
 * {@code ClusteringExecutionStrategy.getReaderProperties} built a fresh property set containing only
 * the spill/memory keys, dropping the impl classes, so the file group reader could not resolve the
 * configured merger. The fix seeds the reader properties from the full write config.
 *
 * <p>Both clustering execution paths call {@code getReaderProperties} and reproduce the bug:
 * the row-writer path ({@code MultipleSparkJobExecutionStrategy#readRecordsForGroupAsRow}, used when
 * {@code hoodie.datasource.write.row.writer.enable} is true — the default the reporter hit) and the
 * RDD path ({@code MultipleSparkJobExecutionStrategy#readRecordsForGroup} when it is false). The test
 * is parameterized over both.
 *
 * <p>This lives in {@code hudi-spark} (not {@code hudi-spark-client}): a CUSTOM/SPARK-typed merger
 * forces the InternalRow write path, which needs a concrete {@code SparkXXXAdapter} and real Spark
 * records — neither is available to {@code hudi-spark-client}'s test classpath / Avro test-data path.
 */
class TestClusteringWithCustomMerger extends HoodieSparkProcedureTestBase {

  test("Test clustering with CUSTOM record merge mode and a custom Spark merger") {
    // hoodie.datasource.write.row.writer.enable: true exercises the row-writer clustering path
    // (the default the reporter hit), false exercises the RDD path. Both call getReaderProperties.
    Seq("true", "false").foreach { rowWriterEnabled =>
      withTempDir { tmp =>
        val tableName = generateTableName
        val basePath = s"${tmp.getCanonicalPath}/$tableName"
        val mergerClass = classOf[CustomSparkRecordMergerForClustering].getName

        spark.sql(
          s"""
             |create table $tableName (
             |  id int,
             |  name string,
             |  price double,
             |  ts long,
             |  part long
             |) using hudi
             | tblproperties (
             |  primaryKey = 'id',
             |  type = 'mor',
             |  orderingFields = 'ts',
             |  "hoodie.write.record.merge.mode" = "CUSTOM",
             |  "hoodie.write.record.merge.strategy.id" = "${HoodieRecordMerger.CUSTOM_MERGE_STRATEGY_UUID}",
             |  "${HoodieReaderConfig.RECORD_MERGE_IMPL_CLASSES_WRITE_CONFIG_KEY}" = "$mergerClass",
             |  "hoodie.parquet.small.file.limit" = "0"
             | )
             | partitioned by (part)
             | location '$basePath'
       """.stripMargin)

        withSQLConf(
          "hoodie.compact.inline" -> "false",
          "hoodie.compact.schedule.inline" -> "false") {
          // Several commits into the same partition create multiple file groups for clustering to
          // combine (small.file.limit=0 keeps every insert in a new base file).
          spark.sql(s"insert into $tableName values (1, 'a1', 10.0, 1000, 100)")
          spark.sql(s"insert into $tableName values (2, 'a2', 20.0, 1001, 100)")
          spark.sql(s"insert into $tableName values (3, 'a3', 30.0, 1002, 100)")
          // An update lands in a log file, so clustering must merge base + log records through the
          // file group reader using the CUSTOM Spark merger.
          spark.sql(s"update $tableName set price = 99.0, ts = 1003 where id = 1")

          // Pin the row-writer setting and propagate the custom merger impl classes (write-side,
          // non-persisted) to the clustering write client. Before HUDI-18980 the reader properties
          // dropped these and clustering failed with "No valid spark merger implementation set".
          val clusteringOptions =
            s"hoodie.datasource.write.row.writer.enable=$rowWriterEnabled," +
              s"${HoodieReaderConfig.RECORD_MERGE_IMPL_CLASSES_WRITE_CONFIG_KEY}=$mergerClass"
          spark.sql(s"call run_clustering(table => '$tableName', options => '$clusteringOptions')").show()

          // Clustering must have produced exactly one completed replace commit.
          val fs = new Path(basePath).getFileSystem(spark.sessionState.newHadoopConf())
          val replaceCommits = HoodieDataSourceHelpers.allCompletedCommitsCompactions(fs, basePath)
            .getInstants.iterator().asScala
            .filter(_.getAction == HoodieTimeline.REPLACE_COMMIT_ACTION)
            .toSeq
          assertResult(1)(replaceCommits.size)

          // All records remain readable after clustering, with the update applied.
          checkAnswer(s"select id, name, price, ts, part from $tableName order by id")(
            Seq(1, "a1", 99.0, 1003, 100),
            Seq(2, "a2", 20.0, 1001, 100),
            Seq(3, "a3", 30.0, 1002, 100)
          )
        }
      }
    }
  }
}

/**
 * A custom Spark record merger whose only purpose is to advertise the CUSTOM merge strategy id, so the
 * table is configured with {@code RecordMergeMode.CUSTOM} and a custom merger impl class. It otherwise
 * reuses the default Spark merge behavior.
 */
class CustomSparkRecordMergerForClustering extends DefaultSparkRecordMerger {
  override def getMergingStrategy: String = HoodieRecordMerger.CUSTOM_MERGE_STRATEGY_UUID
}
